### PR TITLE
pace-util: zarr monitor quantity attribute handling

### PR DESCRIPTION
**Pull request by brianhenn**
## Purpose

Allow the `pace.util.ZarrMonitor` to handle incoming diagnostic `pace.util.Quantity` objects with different dimension orders, without losing the attributes of those Quantities. Also relax the requirements of consistency in incoming zarr monitor Quantities' attributes, beyond consistent units and dimensions. Both of these changes are intended to make the fv3net integration of the zarr monitor more robust. 

## Code changes:

- `pace.util.ZarrMonitor` can now store incoming diagnostic `pace.util.Quantity` objects with different dimension orders and inconsistent attributes (except for units and the set of dimensions, which must be consistent). 

## Checklist
Before submitting this PR, please make sure:

- [X] Docstrings and type hints are added to new and updated routines, as appropriate
- [X] For each public change and fix in `pace-util`, HISTORY has been updated
- [X] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

File: pace-util/pace/util/zarr_monitor.py [hunk 1]
```diff
@@ -168,6 +169,13 @@ def partitioner(self):
     def rank(self):
         return self.comm.Get_rank()
 
+    @property
+    def array_dims(self):
```
Comment (mcgibbon) [inline pace-util/pace/util/zarr_monitor.py:173]: I would suggest making this private `_array_dims`.

File: pace-util/pace/util/zarr_monitor.py [hunk 2]
```diff
@@ -264,16 +273,24 @@ def append(self, quantity):
...
-    def _ensure_compatible_attrs(self, new_quantity):
-        new_attrs = self._get_attrs(new_quantity)
-        if dict(self.array.attrs) != new_attrs:
+    def _get_quantity_dims(self, quantity):
+        return list(self._PREPEND_DIMS + quantity.dims)
+
+    def _transpose(self, quantity):
+        transposed = quantity.transpose(self.array_dims[2:])
+        transposed.update_attrs(dissoc(quantity.attrs, "units"))
+        return transposed
```
Comment (mcgibbon) [inline pace-util/pace/util/zarr_monitor.py:286]: I would suggest deleting this helper, and instead modifying `quantity.transpose` to propagate attrs.

File: pace-util/tests/test_zarr_monitor.py [hunk 3]
```diff
@@ -453,21 +463,57 @@ def test_transposed_diags_write_across_ranks(diag, cube_partitioner, tmpdir_fact
...
+    time_2 = cftime.DatetimeJulian(2010, 6, 20, 6, 15, 0)
+    zarr_monitor_single_rank.store({"time": time_1, "a": diag})
+    new_dims = list(diag.dims)
+    new_dims[-1] = "some_other_dim"
+    diag_2 = pace.util.Quantity(
+        numpy.ones([size + 2 for size in range(len(diag.dims))]),
+        dims=new_dims,
+        units="m",
+    )
+    with pytest.raises(ValueError):
```
Comment (mcgibbon) [inline pace-util/tests/test_zarr_monitor.py:493]: You may want to test that the correct error message is printed, if you care about your more user-friendly error message versus the one in `.transpose`.

File: pace-util/pace/util/quantity.py [hunk 4]
```diff
@@ -532,6 +532,16 @@ def transpose(self, target_dims: Sequence[Union[str, Iterable[str]]]) -> "Quanti
             gt4py_backend=self.gt4py_backend,
         )
 
+    def update_attrs(self, attrs: dict) -> None:
```
Comment (mcgibbon) [inline pace-util/pace/util/quantity.py:535]: Please remove this in favor of propagating the attrs when doing `.transpose`.

**Comment by mcgibbon** [comment]
If the `array_dims` property becomes private and you can remove `update_attrs` in favor of `.transpose` propagating attributes, IMO this is good to merge.

**Review by elynnwu — APPROVED** [review]

**Pull request merged**